src/main/scala/net/snowflake/spark/snowflake/TelemetryReporter.scala

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2015-2020 Snowflake Computing
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.snowflake.spark.snowflake
+
+private[snowflake] object TelemetryReporter {
+
+  private var driverTelemetryReporter: TelemetryReporter = new NoopTelemetryReporter()
+
+  private[snowflake] def setDriverTelemetryReporter(tr: TelemetryReporter): Unit = {
+    driverTelemetryReporter = tr
+  }
+
+  private[snowflake] def resetDriverTelemetryReporter(): Unit = {
+    driverTelemetryReporter = new NoopTelemetryReporter()
+  }
+
+  private val executorTelemetryReporters = new ThreadLocal[TelemetryReporter]() {
+    override protected def initialValue = new NoopTelemetryReporter()
+  }
+
+  private val isExecutorThread = new ThreadLocal[Boolean]() {
```

Comment:
This is not the right way to do this, you should probably use the task context or something to figure out if you are in an executor.

Reply:
Also, why is this a thread local ?

Reply:
This is used determine whether this thread is executor.
If "TaskContext.get() == null" can be used to distinguish driver and executor, this is redundant.